File: 1-two-sum/go/main.go
package main

import "fmt"

// O(n^2) Solution
func twoSum(nums []int, target int) []int {
    if len(nums) < 2 {
        return []int{-1, -1}
    }

    for i := 0; i < len(nums) - 1; i++ {
        for j := i + 1; j < len(nums); j++ {
            if nums[i] + nums[j] == target {
                return []int{i, j}
            }
        }
    }

    return []int{-1,-1}
}

func main() {
	// fmt.Println("Hello, world! ~ GO")

    nums := []int{2,7,11,15}
    target := 9
    fmt.Println("Result = ", twoSum(nums, target))
}

File: 1-two-sum/rust/src/main.rs
use std::collections::HashMap;

fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut map: HashMap<i32, i32> = HashMap::new();
    map.insert(nums[0], 0);
    for index in 1..nums.len() {
        let needed: i32 = target - nums[index];
        if map.contains_key(&needed) {
            return [index.try_into().unwrap(), map[&needed]].to_vec();
        }
        map.insert(nums[index], index.try_into().unwrap());
    }

    return [-1, -1].to_vec();
}

fn main() {
    println!("Hello, world!");

    let nums = Vec::from([2, 7, 11, 15]);
    let target = 9;
    println!("Result is = {:?}", two_sum(nums, target));
}
